Replicated reads and writes using Consistent Hashing

ashwanthkumar opened the pull request:
- CHRing#find(key, n) - returns n nodes from the ring closest to the key - can contain duplicates
- CHRing#findNUnique(key, n) - returns n unique nodes from the right closest to the key
- Changed contract of the Partitioner to return MemberAddress instead of VNode, since VNode is CHRing specific detail
- Added few more tests around CHPartitioner and CHRing

Part of #21

ashwanthkumar committed:
Add multi-node get on CHRing

- CHRing#find(key, n) - returns n nodes from the ring closest to the key - can contain duplicates
- CHRing#findNUnique(key, n) - returns n unique nodes from the right closest to the key
- Changed contract of the Partitioner to return MemberAddress instead of VNode, since VNode is CHRing specific detail
- Added few more tests around CHPartitioner and CHRing

ashwanthkumar committed:
Replacing all instances of bins -> nodes

brewkode commented on src/main/scala/in/ashwanthkumar/suuchi/partitioner/ConsistentHashRing.scala:
```diff
@@ -28,16 +36,75 @@ class ConsistentHashRing(hashFn: Hash, vnodeFactor: Int = 3) {
+      var hash = hashFn.hash(key)
+      (0 until n).map { index =>
+        val (newHash, candidate) = findCandidate(hash)
+        hash = newHash + 1
```
Saw the reference and this as well :)

brewkode commented:
Lets catch up on this before I merge. Hope thats ok.

brewkode committed:
Minor refactoring to get rid of var

ashwanthkumar commented:
@brewkode Shall I go ahead and merge this?

brewkode committed:
First cut - replication router - WARN: not elegant

ashwanthkumar commented on src/test/scala/in/ashwanthkumar/suuchi/router/RouterSpec.scala:
```diff
@@ -11,29 +11,29 @@ class NeverRoute extends RoutingStrategy {
 }
 
-class RouterTest extends FlatSpec {
+class RouterSpec extends FlatSpec {
```
Should we rename this to `HandleOrForwardRouterSpec` ?

ashwanthkumar commented on src/main/scala/in/ashwanthkumar/suuchi/router/HandleOrForwardRouter.scala:
```diff
@@ -27,17 +27,21 @@ class Router(routingStrategy: RoutingStrategy, self: MemberAddress) extends Serv
-            case Some(node) if !node.equals(self) =>
-              log.trace(s"Forwarding request to $node")
+            case items if items.nonEmpty && !items.exists(_.equals(self)) =>
+              val node = items.head
```
Somehow forwarding to just the `items.head` - feels a little weird.

brewkode committed:
Setting things up for replication + changes in Example App

ashwanthkumar committed:
Fixing a small bug in ConsistentHashingRouting where nrReplicas were not picked up

ashwanthkumar committed:
Pluggable Replicator

New Workflow for replication

                                Client

                                  ||
                                  ||
                                  ||
                                  \/

                                Request

                                  ||
                                  ||
                                  ||
                                  \/

                            HandleOrForward

                               /\    ||
[Header:REPLICATION_REQUEST]   ||    ||  Header: ELIGIBLE_NODES=[]
 (times: nrOfReplica)          ||    ||
                               ||    \/

                            ReplicationRouter

HandleOrForward Interceptor always sets a header called "ELIGIBLE_NODES" which contains list of all nodes that're eligible to participate in the current operation as defined by RoutingStrategy.
SequentialReplicator would know to replicate the incomingMessage based on ELIGIBLE_NODES, if REPLICATION_REQUEST header is not set. Now while replicating the message it would add REPLICATION_REQUEST header to those messages so ReplicationRouter on the other node (or process) will not forward the message instead just delegate it to the local handler.

As part of this change we've also made ReplicationRouter an abstract class and provided an implementation called SequentialReplicator which does Sequential replication to all the ELIGIBLE_NODES.

ashwanthkumar committed:
Renaming NodeMarshallers + adding simple happy path tests

ashwanthkumar committed:
Adding simple tests for StringMarshaller

ashwanthkumar committed:
Adding tests for NoNodes found case in HandleOrForwardRouter

ashwanthkumar committed:
Removing some un-used param + minor name changes

ashwanthkumar committed:
Fixing the build on the tests

ashwanthkumar committed:
First cut tests for ReplicationRouter

ashwanthkumar committed:
Removing un-used imports

ashwanthkumar committed:
Fixing the test spec name

ashwanthkumar committed:
Adding tests for SequentialReplicator

ashwanthkumar committed:
Organizing imports across the project

ashwanthkumar committed:
Organising imports across the tests

ashwanthkumar committed:
Merge branch 'master' into multi-find-consistent-hash

ashwanthkumar commented:
Most of my comments are addressed as part of #27

brewkode committed:
Merge pull request #27 from ashwanthkumar/refactor-replicator

Pluggable Replicator

brewkode commented:
@ashwanthkumar Does this need to be merged?

ashwanthkumar commented:
@brewkode #27 was merge into this branch so yeah we need to merge this too.

ashwanthkumar committed:
Reusing the same Metadata.Key instace and moving them to Headers

ashwanthkumar committed:
Removing the TODO comment for header based check

ashwanthkumar committed:
Using full namespace to keep it on par with HandleOrForward

ashwanthkumar committed:
fixing the build - caused due to moving the headers around

ashwanthkumar commented:
@brewkode incorporated the feedback comments from #27 in here. Have one final look at the changes and we can merge it in.

brewkode commented:
@ashwanthkumar Merging this!
